### ISSUE TITLE
Add support for objects with handleEvent in event handlers

### ISSUE THREAD

**Issue by spacejack** [MithrilJS/mithril.js#1939]
## Expected Behavior
When adding event handlers, Mithril should support objects with a `handleEvent` function property in addition to plain functions.

## Current Behavior
Only functions will be used as event handlers

## Possible Solution
Test if the object supplied to an `on[event]` attrs property has a handleEvent function property. Example:

```javascript
{
    handleEvent(e) {/*...*/},
    view() {return m('button', {onclick: this}, 'Click')}
}
```

## Steps to Reproduce (for bugs)
[Demo of lack of support](https://flems.io/#0=N4Ig9grgLgDtIC4oCcIFMA0IDOaA2aAxlGgCaIgB0AVtiFgGYCWBdCA2qAHYCGAtmgqUAFlD556IQmC4lZFEAF8M3foIRVak6bLTyNO7FAAEhYwF5jwADpdjpyLITGADBlv2ZhPE0IBrAAoASisPe2MoYSZsSmkIWQBqBLDlMOEeLlICAFEANz0oALQQmztwwzACSjwwAHMAgHJ0zJz82VMePAJSBqCw+0jo9gaZBuME4zRKKABPGDQAXSK+stSy3KY0AHdg0LL7ZDQoCGQ7PkbptCMG933w84aYG+MGgGFHKGcxicGYuNkgrdwvdGgAjaBQUYYKxeHz+Zy-ZTGawgACCpFIKJWwOxxkUtnxXFsfEofA+AVIYEIEAEskooLApBm0MIKwqVRq9QaZPiJB6KyUKhAvAEQkI2DoWB0cigCkUCywPi4fjYnBAJwkGlEsGwCAA9Hr4jA-LVYmA+Hq+ExIsgWJJZvMKFp5YogA)

## Context
This makes it quite easy to avoid re-creating event handler functions every redraw, and doesn't require binding for `this` context. See [this article](https://medium.com/@WebReflection/dom-handleevent-a-cross-platform-standard-since-year-2000-5bf17287fd38) (pointed out by @JAForbes.)

## Your Environment
Affects all browsers.

**Comment by pygy** [comment]
We should probably support this, yes... I discovered that interface recently probably as a ripple of that article.

It should be noted that the interface only works with `addEventListener`, [not legacy `element.onevent` setters](https://flems.io/#0=N4Ig9grgLgDtIC4oCcIFMA0IDOaA2aAxlGgCaIgB0AVtiFgGYCWBdCA2qAHYCGAtmgqUAFlD556IQmC4lZFEAF8M3foIRVak6bLTyNYAAQBeQ8GE8upAgFEAbnqgAKNAEpgO7GAKU8YAOZOhlDCTNgYhpiRlFAAnjBuiooAOlZghBACspQARmCksZQyhHhMhADWJoZgqaTpmY65+YU8pKT2jgAyYXJoyE4A5HyQuHUA7lwDEWCuSiogvAJChNh0WDpyUAqKALqKQA)... (at least not in Chrome and Firefox). So we'll have to take that into account.

**Comment by spacejack** [comment]
I took an initial run at it [here](https://github.com/spacejack/mithril.js/commit/0a21e8870ee2d5e5a6536e72fb3489fe08d41ccf). Still need to add tests and see if this thing actually works. But a few things I'm unsure of:

* Not sure how a null exception is avoided [here](https://github.com/MithrilJS/mithril.js/blob/next/render/render.js#L582). If it wasn't susceptible before then I don't think it should be now.

* Not sure how `element.onevent` setters were used previously. There are checks to [replace existing ones](https://github.com/MithrilJS/mithril.js/blob/next/render/render.js#L586) in an element but I don't see how they would've been set initially.

**Comment by pygy** [comment]
It looks like a good start... Regarding your questions:

* The `null` check is not needed at that point because the `callback`/`listener` is only assigned when value isn't `null`.

* `key in element` [looks for the presence of `onX` getters/setters](https://flems.io/#0=N4Ig9grgLgDtIC4oCcIFMA0IDOaA2aAxlGgCaIgB0AVtiFgGYCWBdCA2qAHYCGAtmgqUAFlD556IQmC4lZFEAF8M3foIRVak6bLTyNO7GAKU8YAOYAKAOQzCeJoQDW1gARMur0mEIQBsygAjMFIATwBKJRUQXgEhQmw6LB05KAVFAF1FIA), whether or not they've been set previously.

* `onevent` handles `e.redraw = false` when using `mount` and `route`

I'll add a few comments in line...

**Comment by spacejack** [comment]
Commited some [updates](https://github.com/MithrilJS/mithril.js/compare/next...spacejack:handle-event?expand=1), I believe it's now handling element.onevent cases in addition to addEventListener with functions, objects. Hoping it's not adding too much overhead.

Have some working demos running in the browser.

Started adding some tests, however it's failing and I wonder if the domMock dispatchEvent supports EventListener objects?

**Comment by pygy** [comment]
> I wonder if the domMock dispatchEvent supports EventListener objects?

Nope, [it doesn't](https://github.com/MithrilJS/mithril.js/blob/b59ab50ba3c6936ae382fc02af89c0d254614857/test-utils/domMock.js#L281).

**Comment by spacejack** [comment]
Added support for EventListener objects where you pointed out. Added tests which seem to be working now.

I also created a built branch so we can [try it in the browser](https://flems.io/#0=N4Ig9grgLgDtIC4oCcIFMA0IDOaA2aAxlGgCaIgB0AVtiFgGYCWBdCA2qAHYCGAtmgqUAFlD556IQmC4lZFEAF8M3foIRVak6bLTyNO7FAAEhYwF5jwADpdjpyLITGADBlv2ZhPE0IBrAAo0AEorD3tjKGEmbEppCFkAakTw5XDhHi5SAgBRADc9KCDQmzsIwzACSjwwAHMAgHIMrNyC2WNmZDIGjEjo7F7MYzRKKABPGBDw+yiY9gaZBuNE4dGJtABdYtT3MrymNAB3AJLp4y6oCGQ7PkbSJjyes-tbhpge4wb4p0-lvpi4o4oMFdhEIq8AEbQKCLUFgiKleHwgD0yOM2EOPBgxh8Rj0aGQkXW2GGPEIwlMPn8z3hXipfmcs1i3xMAFJjAAmCzmSwuGlIgD8xhOFgAfFYmYCElBkop+fDGf15WkykiGgBBUikBo04JnPVlOVcI22PiUPhAgKkMCECACWSUCFgUhjXqEA1KFQgXgCISEbB0LA6ORQBSKDZYHxcPxsTggK4SDSiWDYBCo5A8Q61JhQQF8ZHYGBktDUMl+ZF8HPCZAsGjYZHNbJoAC0aDaUGbUJYpArVZreDrknGkwoWnDiiAA).

**Comment by spacejack** [comment]
I did find a bug, but it looks like it is a pre-existing issue. [Example](https://flems.io/#0=N4Ig9grgLgDtIC4oCcIFMA0IDOaA2aAxlGgCaIgB0AVtiFgGYCWBdCA2qAHYCGAtmgqUAFlD556IQmC4lZFEAF8M3foIRVak6bLTyNO7FAAEhYwF5jwADpdjpyLITGADBlv2ZhPE0IBrAAo0AEorD3tjKGEmbEppCFkAakTw5XDhHi5SAgBRADc9KCDQmzsIwzACSjwwAHMAgHIMrNyC2WNmZDIGjEjo7F7MYzRKKABPGBDw+yiY9gaZBuNE4dGJtABdYtT3MrymNAB3AJLp4y6oCGQ7PkbSJjyes-tbhpge4wb4p0-lvpi4o4oMFdhEIq8AEbQKCLUFgiKleHwrw+fzOWaxb4mcw41zPJHGAD8-2w+KRzi4EDweDJxjSZSRDQAgqRSA18cEzpyyopbLyuLY+JQ+ECAqQwIQIAJZJQIWBSGNeoRuUoVCBeAIhIRsHQsDo5FAFIoNlgfFw-GxOCArhINKJYNgEAB6J3IHiHWpMKCAvhO7AwHiENDUQN+J18L3CZAsGjYJ3NbJoAC0aDaUCTUJYpHDkejeFjknGkwoWmNiiAA) - Clicking "Add" should increment once, then the event listener should be removed.

I can reproduce this with current Mithril in [this example](https://flems.io/#0=N4IgtglgJlA2CmIBcBWATAOgCwoDQgHsBXAFwAdTkSAnI+fAZ3gQGMT4pkQMArBkfADMICfkgDaoAHYBDMIiTcAFiTCwBIFgSnsdXEAF9c0uQu58NWnfD2KrDEgAIWjgLyPgAHSmPnxHUiOAAy43r4AbhDwAO4AFACUHmG+jtTwJETUPmCxAORQEOG5oT4pvjm5ZMWOuVpEATWOANSOJEoQDBh1OvElZeV5BUV9-b5epaO+DgCeCIHjk-0FDGSwMtOBuRBSsNvwALQARrAELADW1WQyMNsA5ptBGGjwYMEYKC-FyYu+hzLnt2o-igAGECCdqJsAMQg2HVFiZBgESE1MgEbbsajVIhMagAZWY8DYmyk2nguW+kyMlNG2hIxBYSgcMmoJECbQ6XX8TlcvOCNMmAH5HAk3AA+DwczrdEhNJoGAWjQJSIiwWCKxzUib9XIAQRgFO1vnilJNpQVUgt3jAGDA3NiUFORHkOgwhwIUGmuGcZsMAF18LspGcxJIQJl1IoVOQGEgAPRx+pkM63LlgOOQNrUEQaEjTMhmCwGP0GIA). You'll need to use your browser's mobile emulator (touchstart is an event that must be added with `addEventListener`). Count should only increment once and then stop handling touchstart events, however the listener persists.

Finally, using `undefined` instead of `null` also fails to remove an onclick callback: [example](https://flems.io/#0=N4IgtglgJlA2CmIBcBWATAOgCwoDQgHsBXAFwAdTkSAnI+fAZ3gQGMT4pkQMArBkfADMICfkgDaoAHYBDMIiTcAFiTCwBIFgSnsdXEAF9c0uQu58NWnfD2KrDEgAIWjgLyPgAHSmPnxHUiOAAy43r4AbhDwAO4AFACUHmG+jtTwJETUPmCxAORQEOG5oT4pvjm5ZMWOuVpEATWOANSOJEoQDBh1OvElZeV5AEakJNrFyf1JpZO+2iywECwA1oFtHV3+Tq7bwRMzvgD8jgluAHwea53dJE1NBnv7gfVQ8MJSHA8pRp++uQCCMFyn3iexBpXuUgh3jAGDAm1iUAILCI8h0GEGBCgAE9cM4wYYALr4BZSJZiSQgTLqRQqcgMJAAegZ9TISwA5hswAzIG1qCINCQsWQzBYDASDEA)

**Comment by pygy** [comment]
#1804 and related issues are in principle tackled in #1865, I should finish that PR...

**Comment by isiahmeadows** [comment]
Should be fixed now by #1949.
